### PR TITLE
Update docs, re-instate backwards compatability

### PR DESCRIPTION
**Pull request by hexylena**
- Unfortunately we have to retain backwards compatibility, so this patch adds that back in.
- Additionally supports pypi URLs (`#md5=`) so people copy+pasting those directly will get the benefit of hash checking.
- updates docs accordingly

### CODE REVIEW

**Comment by nsoranzo** [comment]
:+1:

**Comment by yhoogstrate** [comment]
Nice @erasche!

I realized that there are a few other commands in which I think it would be desired to have this too:
- `action_type="download_binary"`
- `action_type="download_file"`

And also, within `action_type="setup_r_environment"`, the package tag:

`<action type="setup_r_environment">
    <package>https://github.com/bgruening/download_store/raw/master/DESeq2-1_0_18BiocGenerics_0.6.0.tar.gz6</package>
</action>`

File: lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py [hunk 1]
```diff
@@ -903,26 +927,26 @@ def handle_environment_variables( self, install_environment, tool_dependency, en
         This method works with with a combination of three tool dependency definition tag sets, which are defined
```
Comment (yhoogstrate) [inline lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py]: with with

File: lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py [hunk 2]
```diff
@@ -179,18 +191,34 @@ def url_download( self, install_dir, downloaded_file_name, download_url, extract
...
-            if downloaded_checksum != checksums['md5sum']:
-                raise Exception( 'Given md5 checksum does not match with the one from the downloaded file (%s).' % (downloaded_checksum) )
-        
+
+            # Determine expected value
+            if checksums.has_key('md5sum'):
+                expected = checksums['md5sum']
+            else:
+                if '#md5=' in download_url:
+                    expected = download_url.split('#md5=')[1]
```
Comment (yhoogstrate) [inline lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py]: re.split("#md5[#=]",download_url)

File: lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py [hunk 3]
```diff
@@ -179,18 +191,34 @@ def url_download( self, install_dir, downloaded_file_name, download_url, extract
...
-        
-        if checksums.has_key('md5sum'):
+
+            # Determine expected value
+            if checksums.has_key('sha256sum'):
+                expected = checksums['sha256sum']
+            else:
+                expected = download_url.split('#sha256#')[1]
+
+            if downloaded_checksum != expected:
```
Comment (yhoogstrate) [inline lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py:205]: downloaded_checksum.lower()

File: lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py [hunk 4]
```diff
@@ -179,18 +191,34 @@ def url_download( self, install_dir, downloaded_file_name, download_url, extract
...
+            # Determine expected value
+            if checksums.has_key('md5sum'):
+                expected = checksums['md5sum']
+            else:
+                if '#md5=' in download_url:
+                    expected = download_url.split('#md5=')[1]
+                else:
+                    expected = download_url.split('#md5#')[1]
+
+            if downloaded_checksum != expected:
```
Comment (yhoogstrate) [inline lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py:217]: downloaded_checksum.lower()

File: lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py (hunk 2 above)
Comment (hexylena) [inline lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py]: oooh, better suggestion!

File: lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py (hunk 3 above)
Comment (hexylena) [inline lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py:205]: good catch, no doubt people will have that issue.

**Comment by hexylena** [comment]
Okay, will PR with expansions to include the other `download_` tags.

> And also, within action_type="setup_r_environment", the package tag

there are a whole bunch of these for R/Perl/Etc, shouldn't be too tough to handle them generically.

File: lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py (hunk 3 above)
Comment (natefoo) [inline lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py:205]: `hashlib.hexdigest()` docs do not state whether implementations should return lowercase, although at least for CPython this appears to be the case.

It's probably more likely that `expected` would have uppercase characters. However, it's probably safest to just `.lower()` everything.

**Comment by natefoo** [comment]
I've been working on doing this generically for the last few hours because I thought it'd be great to get it into 15.07 (and the freeze is today) but have just run into a roadblock, as it appears the step objects are destroyed and reinstantiated between the prepare and execute phases (I was attempting to store checksums in the `Download` class after pulling them out of the `Action` element in the prepare method).

The less generic solution will be to copy and paste a line to create the `checksums` dict for every single step class that uses `Download` and then pass that dict into every call of `url_download()` as is already done for the `DownloadByUrl` class. But unless someone wants to redesign the installation recipe manager/step process this is probably the best we are going to do.

File: lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py (hunk 3 above)
Comment (yhoogstrate) [inline lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py:205]: You are absolutely correct (https://docs.python.org/2/library/hashlib.html), and hashlib also has no built-in validation function that takes care of this either. Lowering both would be the cleanest.

**Comment by natefoo** [comment]
- Implemented the line comments from @yhoogstrate in natefoo/galaxy@72b7381fc288fe78026493433171e11d6f7fae76.
- Quasi-generic checksum verification-for-all in natefoo/galaxy@37b231cb3d0ece73daa4ba45149ec14d04abc1d6. Tested `download_*` but not `setup_*_environment`.

**Comment by natefoo** [comment]
@erasche @yhoogstrate As per galaxyproject#444 I pushed my changes to erasche:docs-fix.

**Comment by yhoogstrate** [comment]
Should I merge it in this one? I gave both of you access to my fork in case you need it.

**Comment by nsoranzo** [comment]
@yhoogstrate I'd say yes: https://botbot.me/freenode/galaxyproject/2015-07-14/?msg=44449047&page=1

**Pull request merged**